File: python/leetcode/39-combination-sum.py
class Solution:
    def combinationSum(self, candidates: List[int], target: int) -> List[List[int]]:
        res = set()

        def backtrack(temp, nums, t):
            if t == 0:
                res.add(tuple(sorted(temp.copy())))
                return
            elif t < 0:
                return

            for n in nums:
                temp.append(n)
                backtrack(temp, nums, t-n)
                temp.pop()

        backtrack([], candidates, target)
        return [list(r) for r in res]

File: c#/leetcode/39-combination-sum.cs
public class Solution
{
  public IList<IList<int>> CombinationSum(int[] candidates, int target)
  {
    var res = new HashSet<IList<int>>();

    void Backtrack(List<int> temp, int remaining)
    {
      if (remaining == 0)
      {
        res.Add(new List<int>(temp.OrderBy(x => x).ToList()));
      }
      if (remaining < 0)
      {
        return;
      }

      foreach (var num in candidates)
      {
        temp.Add(num);
        Backtrack(temp, remaining - num);
        temp.RemoveAt(temp.Count - 1);
      }
    }

    Backtrack(new List<int>() { }, target);

    return res.Distinct().ToList();
  }
}